src/spatial/directcrrssa.jl

```diff
@@ -0,0 +1,265 @@
+    if is_hop(p, jump)
+        return accept_hop(p, u, jump)
+    else
+        return accept_rx(p, u, jump)
+    end
+end
+
+function accept_hop(p, u, jump)
+    @unpack hop_rates, spatial_system, rng = p
+    species, site = jump.jidx, jump.src
+    acceptance_threshold = rand(rng) * hop_rate(hop_rates.high, species, site)
+    if hop_rate(hop_rates.low, species, site) > acceptance_threshold
+        return true
+    else
+        # compute the real rate. Could have used hop_rates.high as well.
+        real_rate = evalhoprate(hop_rates.low, u, species, site, spatial_system)
+        return real_rate > acceptance_threshold
+    end
+end
+
+function accept_rx(p, u, jump)
+    @unpack rx_rates, rng = p
+    rx, site = reaction_id_from_jump(p, jump), jump.src
+    acceptance_threshold = rand(rng) * rx_rate(rx_rates.high, rx, site)
+    if rx_rate(rx_rates.low, rx, site) > acceptance_threshold
+        return true
+    else
+        # compute the real rate. Could have used rx_rates.high as well.
+        real_rate = evalrxrate(rx_rates.low, u, rx, site)
+        return real_rate > acceptance_threshold
+    end
+end
+
+"""
+    fill_rates_and_get_times!(aggregation::DirectCRRSSAJumpAggregation, u, t)
+
+reset all stucts, reevaluate all rates, repopulate the priority table
+"""
+function fill_rates_and_get_times!(aggregation::DirectCRRSSAJumpAggregation, integrator, t)
+    @unpack bracket_data, u_low_high, spatial_system, rx_rates, hop_rates, site_rates_high, rt = aggregation
+    u = integrator.u
+    update_u_brackets!(u_low_high::LowHigh, bracket_data, u::AbstractMatrix)
+
+    reset!(rx_rates)
+    reset!(hop_rates)
+    fill!(site_rates_high, zero(eltype(site_rates_high)))
+
+    rxs = 1:num_rxs(rx_rates.low)
+    species = 1:(aggregation.numspecies)
+
+    for site in 1:num_sites(spatial_system)
+        update_rx_rates!(rx_rates, rxs, u_low_high, integrator, site)
+        update_hop_rates!(hop_rates, species, u_low_high, site, spatial_system)
+        site_rates_high[site] = total_site_rate(rx_rates.high, hop_rates.high, site)
+    end
+
+    # setup PriorityTable
+    reset!(rt)
+    for (pid, priority) in enumerate(site_rates_high)
+        insert!(rt, pid, priority)
+    end
+    nothing
+end
+
+"""
+    update_dependent_rates!(p, integrator, t)
+
+recalculate jump rates for jumps that depend on the just executed jump (p.prev_jump)
+"""
+function update_dependent_rates!(p::DirectCRRSSAJumpAggregation, integrator, t)
+    jump = p.prev_jump
+    if is_hop(p, jump)
+        update_brackets!(p, integrator, jump.jidx, (jump.src, jump.dst))
+    else
+        update_brackets!(p, integrator, p.jumptovars_map[reaction_id_from_jump(p, jump)], jump.src)
+    end
+end
+
+function update_brackets!(p, integrator, species_to_update, sites_to_update)
+    @unpack rx_rates, hop_rates, site_rates_high, u_low_high, bracket_data, vartojumps_map, spatial_system = p
+    u = integrator.u
+    for site in sites_to_update, species in species_to_update
```

Comment:
I always forget, but make sure this is the correct memory order to update over.

Reply:
`u_low_high[species, site]` indicates that `species` should be in the inner-most loop. So this is correct.